prov/shm/src/smr_ep.c

```diff
@@ -313,90 +340,189 @@ static void smr_progress_iov(struct smr_cmd *cmd, struct smr_ep_entry *entry,
+		break;
+	case smr_src_iov:
+		err = smr_progress_iov(cmd, entry->iov, entry->iov_count,
+				       &total_len, ep, 0);
+		break;
+	default:
+		FI_WARN(&smr_prov, FI_LOG_EP_CTRL,
+			"unidentified operation type\n");
+		err = -FI_EINVAL;
+	}
+	ret = ep->rx_comp(ep, entry->context, entry->flags, total_len,
+			  entry->iov[0].iov_base, &addr, cmd->hdr.op.tag, err);
+	if (ret) {
+		FI_WARN(&smr_prov, FI_LOG_EP_CTRL,
+			"unable to process rx completion\n");
+	}
+	freestack_push(ep->recv_fs, entry);
 
-		dlist_entry = dlist_remove_first_match(&recv_queue->recv_list,
-						       recv_queue->match_recv,
-						       &match_attr);
-		if (!dlist_entry) {
-			if (freestack_isempty(ep->unexp_fs))
-				break;
-			unexp = freestack_pop(ep->unexp_fs);
-			memcpy(&unexp->cmd, cmd, sizeof(*cmd));
-			ofi_cirque_discard(smr_cmd_queue(ep->region));
-			dlist_insert_tail(&unexp->entry, &ep->unexp_queue.msg_list);
-			continue;
-		}
-		entry = container_of(dlist_entry, struct smr_ep_entry, entry);
+	return ret;
+}
+
+static int smr_progress_cmd_rma(struct smr_ep *ep, struct smr_cmd *cmd)
+{
+	struct smr_domain *domain;
+	struct iovec iov[SMR_RMA_IOV_LIMIT];
+	size_t iov_count;
+	size_t total_len = 0;
+	int ret = 0;
 
-		switch (cmd->hdr.op.op_src){
-		case smr_src_inline:
-			smr_progress_inline(cmd, entry);
+	domain = container_of(ep->util_ep.domain, struct smr_domain,
+			      util_domain);
+
+	for (iov_count = 0; iov_count < cmd->hdr.op.iov_count; iov_count++) {
+		ret = smr_mr_verify(domain, cmd->data.rma_iov[iov_count].len,
+				    cmd->data.rma_iov[iov_count].addr,
+				    cmd->data.rma_iov[iov_count].key,
+				    cmd->hdr.op.op == ofi_op_write ?
+				    FI_REMOTE_WRITE : FI_REMOTE_READ);
+		if (ret)
 			break;
-		case smr_src_inject:
-			smr_progress_inject(cmd, entry, ep);
+
+		iov[iov_count].iov_base = (void *) cmd->data.rma_iov[iov_count].addr;
+		iov[iov_count].iov_len = cmd->data.rma_iov[iov_count].len;
+	}
+
+	ofi_cirque_discard(smr_cmd_queue(ep->region));
+	cmd = ofi_cirque_head(smr_cmd_queue(ep->region));
+	if (cmd->hdr.op.op != ofi_op_write &&
+	    cmd->hdr.op.op != ofi_op_read_req)
+		return -FI_EINVAL;
+
+	switch (cmd->hdr.op.op_src) {
+	case smr_src_inline:
+		ret = smr_progress_inline(cmd, iov, iov_count, &total_len, ret);
+		break;
+	case smr_src_inject:
+		ret = smr_progress_inject(cmd, iov, iov_count, &total_len, ep, ret);
+		break;
+	case smr_src_iov:
+		ret = smr_progress_iov(cmd, iov, iov_count, &total_len, ep, ret);
+		break;
+	default:
+		FI_WARN(&smr_prov, FI_LOG_EP_CTRL,
+			"unidentified operation type\n");
+		ret = -FI_EINVAL;
+	}
+
+	return ret;
+}
+
+void smr_progress_cmd(struct smr_ep *ep)
+{
+	struct smr_cmd *cmd;
+	int ret;
+
+	fastlock_acquire(&ep->region->lock);
+	fastlock_acquire(&ep->util_ep.rx_cq->cq_lock);
+
+	while (!ofi_cirque_isempty(smr_cmd_queue(ep->region))) {
+		cmd = ofi_cirque_head(smr_cmd_queue(ep->region));
+
+		switch (cmd->hdr.op.op) {
+		case ofi_op_msg:
+			/* fall through */
```

Comment:
nit: you can put two case statements together without needing the fall through comment.